### DISCUSSION TITLE
mDNS response is not sent out over STM32U575zi-WiFi

### DISCUSSION THREAD

**Discussion by VineetaNarkhede-eaton** [zephyrproject-rtos/zephyr#71540] [Q&A]
Hi,

I am facing an issue in running mDNS over STM32U575zi on WiFi interface. Device (Dev board) is in AP mode and connected with Laptop over WiFi. Enabled the mDNS responder in zephyr. Hostname configured is "zephyr"
When I send mdns query from Laptop i.e zephyr.local, the query is successfully reaching the device from the laptop. However, the response from the device does not appear to be reaching the laptop. I have verified that the response is reaching till WiFi driver and the response data is valid. Wireshark logs on the laptop shows query but no response visible. Interestingly, mDNS query-response works sometimes suddenly(very rare) without any code changes. 

I tried the solution provided on this issue https://github.com/zephyrproject-rtos/zephyr/discussions/48327, but no improvement in the result.

**Setup**:
Dev board: NUCLEO-U575ZI-Q
WiFi chip:1YN-CY43439
zephyr OS : v3.2.0
infineon_zephyr_connectivity commit id : 272e2b8da908c7d796ec62757cc8abbe03a91e4c
 
I would appreciate any help with this matter.

**Comment by jukkar** [comment]
Your Zephyr version is very old and currently not supported. Can you upgrade to latest version and/or to upstream and replicate the issue?

**Reply by VineetaNarkhede-eaton** [reply]
Okay, currently we are in a process of upgrade to zephyr 3.6.0.

**Reply by VineetaNarkhede-eaton** [reply]
Hi @jukkar 

I enabled the mDNS in zephyr 3.6.0, but getting bus fault in ethernet stack in ethernet.c, function name: ethernet_fill_header().
![image](https://github.com/zephyrproject-rtos/zephyr/assets/88422239/3242a62b-e3bb-4ae9-aef8-1cc2fd4d7e5e)

Bus fault is coming because net_pkt_lladdr_src(pkt)->addr is null pointer, ethernet_fill_header() is called through setup_ipv4_addr() from mdns_responder.c
![image](https://github.com/zephyrproject-rtos/zephyr/assets/88422239/53c60431-4da4-4f52-bdaa-d0f7fc1acc5a)

If I call mdns_responder_init() explicitly from main function then bus fault is not coming.

Could you please provide some insights here.
Should I convert this discussion to bug?

**Comment by rajolegajendra-eaton** [comment]
We are initializing WiFi in Application. mDNS responder listens to 224.0.0.251 over port 5353. So basically this IP and Port needs to be binded to network context. So for IPV4 `net_ipv4_igmp_join()` is used which creates packet and transmits it. `net_ipv4_igmp_join()` is getting called from two locations:

1. `init_listener()` : This function is directly calling `net_ipv4_igmp_join`.
2. `mdns_iface_event_handler()` : This callback is registered with `net_mgmt_init_event_callback()` and in this callback `net_ipv4_igmp_join` is getting called only if `NET_EVENT_IF_UP` event is generated. That means complete network layer status is checked prior to calling `net_ipv4_igmp_join`.

While going through `net_ipv4_igmp_join` call sequence found that while transmitting packet, zephyr 3.6 is just checking if `NET_IF_LOWER_UP` flag is set or not. If this flag is setting then it is transmitting the packet. Since WiFi initialization has not happened and even L2 layer is not up we are getting BUS Fault for accessing L2 Source Address since it is nullptr. When we checked in Zephyr 3.2 found that to transmit packet it is checking whether `NET_IF_UP` flag is set or not, that mean packets will be transmitted only if complete network layer is up. So changing this flag to `NET_IF_UP` resolved BUS fault error and mDNS was working Zephyr 3.6 with U5 WiFi setup.

![image](https://github.com/zephyrproject-rtos/zephyr/assets/95759451/508886f3-b2a5-4874-b881-be1c4b058faa)

Found one commit where `NET_IF_UP` is changed to `NET_IF_LOWER_UP`.
Link: https://github.com/zephyrproject-rtos/zephyr/commit/89b32a0f5db56851fd178acbcbba8d625b5f42b3

**Reply by jukkar** [reply]
Could you create an issue about this, the discussions is not a proper place for this.

**Reply by VineetaNarkhede-eaton** [reply]
Sure, Converting this discussion into issue.
